Integrate Via's new DB with Pyramid and pytest

seanh opened the pull request:
Depends on <https://github.com/hypothesis/via/pull/1077/>. This PR adds all the manual parts of adding that're needed after running `make template`. This is all the stuff that https://github.com/hypothesis/cookiecutters/pull/128 would have done for Via automatically if we had merged that PR.

## Testing

There's not much to test here since there aren't any models or tables yet. You can try out the various `make` commands. https://github.com/hypothesis/via/pull/1072 actually adds some code that you can test.

robertknight commented on .github/workflows/ci.yml:
```diff
@@ -62,6 +63,11 @@ jobs:
     runs-on: ubuntu-latest
+    services:
+      postgres:
+        image: postgres:11.5-alpine
```
Are we planning to use the current Postgres release (15.x) for this project or stick with the same Pg version as other projects?

seanh replied:
No I think we'll start Via off on the latest Pg version, meaning I'll need to change the version number here (this PR is in a very draft state still)

seanh committed:
Integrate Via's new DB with Pyramid and pytest

seanh commented on tests/unit/conftest.py:
```diff
@@ -34,8 +35,10 @@ def pyramid_config(pyramid_settings):
 ):
     pyramid_request = testing.DummyRequest()
     apply_request_extensions(pyramid_request)
+    pyramid_request.db = db_session
```
Add `pyramid_request.db` in the tests

seanh commented on tests/unit/via/app_test.py:
```diff
@@ -43,29 +42,6 @@ def test_settings_are_configured_from_environment_variables(pyramid_settings):
         assert settings[key] == value
 
 
-@pytest.mark.usefixtures("os")
-def test_app(configurator, pyramid, pyramid_settings):
-    create_app()
-
-    expected_settings = dict(
-        {"h_pyramid_sentry.filters": SENTRY_FILTERS}, **pyramid_settings
-    )
-    expected_settings["data_directory"] = Path(pyramid_settings["data_directory"])
-
-    pyramid.config.Configurator.assert_called_once_with(settings=expected_settings)
-    configurator.make_wsgi_app.assert_called_once_with()
-
-
-@pytest.fixture
-def configurator(pyramid):
-    return pyramid.config.Configurator.return_value
-
-
-@pytest.fixture(autouse=True)
-def pyramid(patch):
-    return patch("via.app.pyramid")
```
This test was just annoying, using mocks instead of using the `pyramid.testing` stuff properly. With this PR's changes to `app.py` the test now crashes with:

    AttributeError: Mock object has no attribute 'registry'

You don't want to make things like `pyramid` or Pyramid's `Configurator` or `request` etc, you want to use the [pyramid.testing](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/testing.html) stuff. I really don't think there's much value in testing the `create_app()` function: there's no logic in it, and if it's broken you'll know pretty fast. So I've just added a coverage ignore to it

seanh commented on .cookiecutter/includes/requirements/prod.in:
```diff
@@ -8,9 +8,11 @@ pyramid-exclog
 pyramid-jinja2
 pyramid-sanity
 pyramid-services
+pyramid-tm
 requests
 whitenoise
 google-auth-oauthlib
 marshmallow
 webargs
 youtube-transcript-api
+zope.sqlalchemy
```
Adding the two requirements needed to integrate SQLAlchemy and Pyramid

seanh commented on tests/conftest.py:
```diff
@@ -46,3 +57,45 @@ def httpretty_():
+@pytest.fixture(scope="session")
+def db_engine():
+    engine = create_engine(environ["DATABASE_URL"])
+
+    Base.metadata.drop_all(engine)
+    Base.metadata.create_all(engine)
+
+    return engine
+
+
+@pytest.fixture(scope="session")
+def db_sessionfactory():
+    return sessionmaker()
+
+
+@pytest.fixture
+def db_session(db_engine, db_sessionfactory):
+    """Return the SQLAlchemy database session.
+
+    This returns a session that is wrapped in an external transaction that is
+    rolled back after each test, so tests can't make database changes that
+    affect later tests.  Even if the test (or the code under test) calls
+    session.commit() this won't touch the external transaction.
+
+    This is the same technique as used in SQLAlchemy's own CI:
+    https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites
+    """
+    connection = db_engine.connect()
+    transaction = connection.begin()
+    session = db_sessionfactory(
+        bind=connection, join_transaction_mode="create_savepoint"
+    )
+    set_factoryboy_sqlalchemy_session(session)
+
+    yield session
+
+    clear_factoryboy_sqlalchemy_session()
+    session.close()
+    transaction.rollback()
+    connection.close()
```
These are the three fixtures that you need to get the DB working in the unit and functional tests. Of course the `db_session` fixture is the only one that the tests actually need to use, `db_engine` and `db_sessionfactory` are just there because `db_session` needs them (and they're separate fixtures because they only need to be executed once per test run, whereas `db_session` is executed once per test)

seanh commented on tests/factories/factoryboy_sqlalchemy_session.py:
```diff
@@ -0,0 +1,48 @@
+    #
+    #     class MyFactory:
+    #         class Meta:
+    #             sqlalchemy_session = session
+    #
+    # Because we don't have `session` available to us at import time.
+    # So we have to do it this way instead.
+    #
+    # See:
+    # https://factoryboy.readthedocs.io/en/latest/orms.html#sqlalchemy
+    # https://factoryboy.readthedocs.io/en/latest/reference.html#factory.Factory._meta
+    for factory_class in _sqlalchemy_factory_classes():
+        # pylint:disable=protected-access
+        factory_class._meta.sqlalchemy_session = session
+
+        if persistence:
+            factory_class._meta.sqlalchemy_session_persistence = persistence
+
+
+def clear_factoryboy_sqlalchemy_session():
+    # Delete the sqlalchemy session from all our test factories.
+    # Just in case, so we don't have references to the session hanging about.
+    for factory_class in _sqlalchemy_factory_classes():
+        factory_class._meta.sqlalchemy_session = None  # pylint:disable=protected-access
+
+
+def _sqlalchemy_factory_classes():
+    """Return all the SQLAlchemy factory classes from tests.factories."""
+
+    # Get the package that this module belongs to.
+    package = sys.modules[sys.modules[__name__].__package__]
+
+    for value in package.__dict__.values():
+        try:
+            is_sqla_factory = issubclass(value, SQLAlchemyModelFactory)
+        except TypeError:
+            is_sqla_factory = False
+
+        if is_sqla_factory:
+            yield value
```
This is a necessary hack that we've had for years that I've copy-pasted exactly from other projects that hooks `factory_boy` up to your test SQLAlchemy session.

The functions are imported and called by `tests/conftest.py` and by `pshell.py`.

seanh commented on via/db.py:
```diff
@@ -0,0 +1,57 @@
+from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass, Session
+
+
+class Base(MappedAsDataclass, DeclarativeBase):
```
`MappedAsDataclass` enables SQLAlchemy 2's dataclasses support (so Via's SQLAlchemy model classes are dataclasses), which is really great. For example all the models get a good `__repr__()` for free, they work with `dataclasses.asdict()` and `dataclasses.astuple()`, etc etc.

seanh commented on via/db.py:
```diff
@@ -0,0 +1,57 @@
+            pass
+
+    if not stamped:
+        alembic.command.stamp(alembic.config.Config("conf/alembic.ini"), "head")
```
This is copy-pasted from other projects and used in dev only (so I don't think it's worth unit testing), it just automatically stamps the DB with the latest Alembic revision in dev, so that people don't have to stamp it manually (which is a PITA)

seanh commented on tests/unit/conftest.py:
```diff
@@ -1,6 +1,7 @@
 """A place to put fixture functions that are useful application-wide."""
 import functools
 
+import pytest
```
Hokay. This file already contained an `@pytest.fixture` but it was not importing `pytest`, and yet it worked? :raised_eyebrow: There's an `import *` in this file that I suspect must have been importing `pytest`. Best to have it here explicitly I think (and we should fix that `import *`, but that's a separate issue)

seanh commented on via/db.py:
```diff
@@ -0,0 +1,57 @@
+    if config.registry.settings.get("dev"):
+        # Create the database tables if they don't already exist.
+        Base.metadata.create_all(engine)
+        stamp(engine)
```
In dev if you add a new model class and then just run `make dev` this will automatically create the DB table for you. But we don't want SQLAlchemy to do that in prod: in prod we want to use DB migrations

seanh commented on via/models/__init__.py:
```diff
@@ -0,0 +1,2 @@
+def includeme(_config):  # pragma: no cover
+    pass
```
This needs to be here so that `app.py` can `config.include("via.models")`. The next PR (https://github.com/hypothesis/via/pull/1072) will actually add a model here

seanh commented on via/pshell.py:
```diff
@@ -1,2 +1,21 @@
+)
+
+
+def setup(env):
+    env["tm"] = env["request"].tm
+    env["tm"].__doc__ = "Active transaction manager (a transaction is already begun)."
+    env["request"].tm.begin()
+
+    env["db"] = env["request"].db
+    env["db"].__doc__ = "Active DB session."
+
+    env["m"] = env["models"] = models
+    env["m"].__doc__ = "The via.models package."
+
+    env["f"] = env["factories"] = factories
+    env["f"].__doc__ = "The test factories for quickly creating objects."
+    set_factoryboy_sqlalchemy_session(env["request"].db)
```
This just makes the models, DB session and factories convenient to use from `make shell`. See the testing instructions in https://github.com/hypothesis/via/pull/1072 for an example.

I wonder about also pre-importing `sqlalchemy.select` for convenience here as well

marcospri commented on via/db.py:
```diff
@@ -0,0 +1,57 @@
+
+    if config.registry.settings.get("dev"):
+        # Create the database tables if they don't already exist.
+        Base.metadata.create_all(engine)
```
This needs to take alembic into account, see the version over LMS:


https://github.com/hypothesis/lms/blob/main/lms/db/__init__.py#L114


In the current version if using alembic to run migrations `make dev` will try to create the tables again and fail to start.

We need to only use `create_all` when the alambic table/row is not there and run the migrations otherwise.

seanh replied:
I think this also needs to be removed from the web worker process. Currently there's a problem that Via has 4 gunicorn workers in dev and when you run `make dev` they all hit this line of code at the same time and all try to create the new tables at once.

My thought is to have the normal `gunicorn` processes *not* call `create_all()` and to have a separate process that _just_ calls `create_all()` and then exits. I'll add this separate process to `supervisord-dev.conf` alongside the existing `gunicorn` and other processes, so when you run `make dev` this new process will also get run and create the tables (if necessary) and then exit.

We already have a `bin/make_db` script that might be all we need for this new process...

seanh replied:
> This needs to take alembic into account, see the version over LMS:
> 
> https://github.com/hypothesis/lms/blob/main/lms/db/__init__.py#L114
> 
> In the current version if using alembic to run migrations `make dev` will try to create the tables again and fail to start.
> 
> We need to only use `create_all` when the alambic table/row is not there and run the migrations otherwise.

I don't think this is right. `create_all()` only creates those tables that don't already exist. It ignores tables that do already exist. That's why I can run `make db` repeatedly and it succeeds each time. The version in LMS looks incorrect to me.

marcospri replied:
>  Via has 4 gunicorn workers in dev and when you run make dev they all hit this line of code at the same time 

You are right, I was probably hitting this while testing migrations here.

I still think the version of this over LMS is superior, it allows people that don't care about migrations to just `make db` to get their environment up to date.

The the version here (and the previous version LMS) will fail every time a migration does something that was not adding a new table. Eg adding new column will break the the app util dev either:

- destroy the DB and `make db` again
- manually remember to run migrations.

seanh commented:
Rebased...

seanh committed:
Move dev DB creation to a separate process

Context
-------

At startup (when you run `make dev`) Via's Gunicorn processes attempt to
create the DB tables if they don't already exist. This is done in dev
only, not in production:

    # via/db.py

    def includeme(config):
        ...

        if config.registry.settings.get("dev"):
            # Create the database tables if they don't already exist.
            Base.metadata.create_all(engine)

Problem
-------

Via actually has four Gunicorn workers in dev (see the `workers = 4` in
`conf/gunicorn/dev.conf.py`). All 4 try to create the DB tables at the
same time and 3 of them crash, printing out error messages.

The issue is kind of difficult to reproduce:

1. You need to actually have a model, so you won't be able to produce
   the issue on the `db` branch without changes. See
   https://github.com/hypothesis/via/pull/1072 and
   https://github.com/hypothesis/via/pull/1093 for branches on which you
   can reproduce the issue.

2. The model's table needs to actually be missing from the DB table. You
   can do that by running this command:

       make services args=down && make services && make dev

3. Since the issue is a race condition you won't be able to reproduce it
   reliably, but if you run the above command several times you should
   see it.

Solution
--------

This commit moves the `sqlalchemy.create_all()` call out of the Gunicorn
worker process and instead adds a separate `make_db` process to the list
of processes that `supervisord-dev.conf` runs when you run `make dev`.
This new `make_db` process starts up, calls `create_all()`, and then
exits, and since it's a singular process there's no collision.

seanh commented:
Rebased

marcospri commented on bin/make_db:
```diff
@@ -1,5 +1,35 @@
+from via.db import Base, create_engine
+
+
+def is_stamped(engine) -> bool:
```
This function is more clear than the same thing but without a name we have in other projects :+1:
